internal/ingest/linksystem.go

```diff
@@ -182,82 +182,76 @@ func (ing *Ingester) ingestEntryChunk(publisher peer.ID, adCid cid.Cid, entryChu
 	return nil
 }
 
-// ingestAd fetches all the entries for a single advertisement
-func (ing *Ingester) ingestAd(publisher peer.ID, adCid cid.Cid) error {
+type adIngestError struct {
+	state adIngestState
```

Comment:
Instead of having different error states, would it be simpler to use a bool `permanent bool` instead?

Reply:
This is on purpose for a couple reasons:

1. It makes adding metrics to these error states more useful. We could have an ingest ad error counter metric and we add a label for each kind of error. Then we know something is wrong if we suddenly see a spike in say decoding errors. And, more helpfully, we know the general kind of thing that is wrong.
2. The caller should decide what to do in the different error states. While we could add a `permanent` bool as well, the caller may want to do something else depending on the error state. For example, in a datastore err we may want to treat that as a temporary error if our datastore is actually a network filesystem rather than a local filesystem. ingestAd shouldn't try to predict what the caller would want to do in it's different errors. It should just report what failed.

Thoughts?

Reply:
see https://github.com/filecoin-project/storetheindex/pull/239/files#diff-50d2db719b550ec2183b831d2727d9184635a5647247ea8872ff82d91b6e31bdR730 for 1.

Reply:
OK, then there is a code-oriented reason for the states, so keep it.